deserialize/__init__.py

```diff
@@ -53,14 +53,14 @@ def _deserialize(class_reference, data, debug_name, throw_on_unhandled: bool):
 
     if is_union(class_reference):
         valid_types = union_types(class_reference)
+        last_exception = None
         for valid_type in valid_types:
             try:
                 return _deserialize(valid_type, data, debug_name, throw_on_unhandled)
-            except DeserializeException:
-                pass
-        raise DeserializeException(
-            f"Cannot deserialize '{type(data)}' to '{class_reference}' for '{debug_name}'"
-        )
+            except DeserializeException as exception:
+                exception.__context__ = last_exception
+                last_exception = exception
+        raise last_exception
```

Comment:
Potential solution for your concernes voiced in [this comment](https://github.com/dalemyers/deserialize/pull/15#issuecomment-555430402)
``` python
raise DeserializeException(
            f"Cannot deserialize '{type(data)}' to '{class_reference}' for '{debug_name}'"
        ) from last_exception
```

Reply:
I've just pushed a change to `develop`. Can you try it and see what you think?

Reply:
Didn't change a thing for [the code example above](https://github.com/dalemyers/deserialize/pull/15#issue-342585939). It still doesn't tell me where exactly the error is.
I don't treat Optional as special case here, it's just the Union I use in my example.

Reply:
Ah, yeah, sorry. I was testing a slightly different case. I can't really look more just now, but I'll try and investigate later this week.

Reply:
Just to be clear: My issue with this PR stems from the fact that we are moving from an error which isn't particularly helpful, to one which is misleading in the majority of cases. I think that's a worse position to be in. I'd rather make the error more helpful in all cases. If we can't, I'd prefer it to be slightly unhelpful than misleading.